### PR TITLE
feat!: add Watchable interface

### PR DESCRIPTION
**Pull request by letian0805**
fix: typo
feat: add Watchable
#### Description (what this PR does / why we need it):

Config source has a method "Watch", but some sources cannot be watched and can only be loaded.So it is necessary to add a Watchable interface.
#### Which issue(s) this PR fixes:

fixes #

### CODE REVIEW

**Comment by ymh199478** [comment]
For some sources, watch can do nothing.  in other words watch has no behavior.

Is there any other reason do you need to distinguish between them (loader and source)?

**Comment by kagaya85** [comment]
Separate them seems to provide no visible benefit but rather more complexity, I think it makes the user confused about when to use `NewSource` or `NewLoader`. For the same function (load config), we want to provide as an simple interface as possible and should not make users waste time on how to choose them.

**Comment by letian0805** [comment]
> For some sources, watch does nothing. in other words watch has no behavior.
> 
> Is there any other reason do you need to distinguish between them (loader and source)?

User is developer. They may load configure from other sources, those don't need watch. If use Source, they need implement empty Watch interface. This looks even more confusing. Analogous to the io package in the Go standard library, which provides both io.Closer and io.ReadCloser, this is the reason.

**Comment by letian0805** [comment]
> Separate them seems to provide no visible benefit but rather more complexity, I think it makes the user confused about when to use `NewSource` or `NewLoader`. For the same function (load config), we want to provide as an simple interface as possible and should not make users waste time on how to choose them.

Analogous to the io package in the Go standard library, which provides both io.Closer and io.ReadCloser, this is the reason.

File: config/config.go [hunk 1]
```diff
@@ -107,13 +107,15 @@ func (c *config) Load() error {
 			c.log.Errorf("failed to merge config source: %v", err)
 			return err
 		}
-		w, err := src.Watch()
-		if err != nil {
-			c.log.Errorf("failed to watch config source: %v", err)
-			return err
+		if s, ok := src.(Source); ok {
```
Comment (kagaya85) [inline config/config.go:110]: Perhaps we can take a discussion of the specific interface, for a config source, `Load()` method is necessary to implement however `Watch()` is not, so the `Watch()` method can be independent from `Source` interface, such as an interface named `Watchable`, and keep the `NewSource` function to maintain the existing usage unchanged. 

```go
type Source interface {
    Load() ([]*KeyValue, error)
}

type Watchable interface {
    Watch() (Watcher, error)
}
```

we can assert here whether config source implements the `Watchable` interface, we'd better not add a new entry function like `NewLoader`, what do you think about it?

Comment (letian0805) [reply config/config.go:110]: Yes, you are right. This is better.

File: config/env/env.go [hunk 2]
```diff
@@ -7,30 +7,38 @@ import (
...
+var _ config.Source = (*envSource)(nil)
+
 type env struct {
-	prefixs []string
+	prefixes []string
+}
+
+type envSource struct {
+	*env
 }
```
Comment (kagaya85) [inline config/env/env.go:18]: Is `envSource` necessary?

File: config/env/env_test.go [hunk 3]
```diff
@@ -367,7 +367,7 @@ func Test_env_load(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			e := &env{
-				prefixs: tt.fields.prefixs,
+				prefixes: tt.fields.prefixs,
```
Comment (kagaya85) [inline config/env/env_test.go:370]: also use `tt.fields.prefixes`?

File: config/options.go [hunk 4]
```diff
@@ -28,7 +28,7 @@ type options struct {
 // WithSource with config source.
 func WithSource(s ...Source) Option {
 	return func(o *options) {
-		o.sources = s
+		o.sources = append(o.sources, s...)
```
Comment (kagaya85) [inline config/options.go:31]: Maybe the original `WithSource` is enough？

**Comment by kagaya85** [comment]
PTAL @tonybase

File: config/env/env.go (hunk 2 above)
Comment (letian0805) [reply config/env/env.go]: Removed

File: config/env/env_test.go (hunk 3 above)
Comment (letian0805) [reply config/env/env_test.go]: Updated

File: config/options.go (hunk 4 above)
Comment (letian0805) [reply config/options.go:31]: User maybe use WithSource many times with different Source

Comment (kagaya85) [reply config/options.go:31]: `WithSource` can accept multiple parameters, what are the scenarios that require the use of multiple `WithSource`?

Comment (letian0805) [reply config/options.go:31]: User don't know WithSource can only use once with all sources, they maybe use twice and more with different sources. The framework should be comfortable for users to use.

**Comment by kagaya85** [comment]
Thanks for contribution, basically LGTM.

**Review by ymh199478 — APPROVED** [review]
LGTM，there is a potential **BC**, see if it needs to be marked

**Comment by letian0805** [comment]
@tonybase  Please review this PR.

**Comment by tonybase** [comment]
There is a BC and splitting the interface is not recommended.

**Pull request closed without merging**